Enable more Go linters: varcheck,bodyclose,structcheck

awly opened the pull request:
These linters required sprinkling some `//nolint` comments due to false
positives. All changes should be noop, except for
`integration/integration_test.go`.

The integration test was ignoring `recordingMode` test case parameter
and always used `RecordAtNode`. When switching to `recordingMode`, test
cases with `RecordAtProxy` fail with a confusing error about missing
user agent. Filed https://github.com/gravitational/teleport/issues/3606
to track that separately and unblock enabling `structcheck` linter.

Updates #3551

Commit:
Enable more Go linters: varcheck,bodyclose,structcheck

All changes should be noop, except for
`integration/integration_test.go`.

The integration test was ignoring `recordingMode` test case parameter
and always used `RecordAtNode`. When switching to `recordingMode`, test
cases with `RecordAtProxy` fail with a confusing error about missing
user agent. Filed https://github.com/gravitational/teleport/issues/3606
to track that separately and unblock enabling `structcheck` linter.

russjones commented on integration/integration_test.go:
```diff
@@ -862,6 +862,11 @@ func (s *IntSuite) TestDisconnectScenarios(c *check.C) {
 		},
 	}
 	for _, tc := range testCases {
+		// TODO: fix the test to work with RecordAtProxy
+		if tc.recordingMode == services.RecordAtProxy {
+			c.Log("Skipping test case with recoridngMode=services.RecodtAtProxy due to https://github.com/gravitational/teleport/issues/3606")
+			continue
+		}
```
We should fix this test instead of adding a log line and TODO to fix it before we merge this PR. Otherwise we'll forget and never come back to it.

awly replied:
I couldn't figure it out. If you have a quick idea, i'd be happy to take another stab.
I filed a bug, since this is a pre-existing problem that might take some effort to fix.

awly replied:
OK, after some digging it turned out trivial - the test didn't enable agent forwarding on the proxy.
Without agent forwarding, proxy recording mode can't work.

russjones commented on lib/web/ui/cluster.go:
```diff
@@ -46,7 +46,7 @@ type Cluster struct {
 }
 
-//nolint:unused,deadcode
+//nolint:unused,deadcode,varcheck
```
If this is dead code, let's just remove it.

awly replied:
Done.
Wasn't sure what the policy on these global loggers was.

russjones reviewed (commented):
Not sure how I feel about the false positives for these linters. The `//nolint` directives feel wrong to me.

@fspmarshall thoughts?

awly reviewed (commented):
> Not sure how I feel about the false positives for these linters. The //nolint directives feel wrong to me.

The `bodyclose` ones aren't actually false positive, they're just a side-effect of using asserts
What happens is: when we make an HTTP request and expect it to fail but it succeeds, the body will remain unclosed while the other tests run.
If we migrated away from asserts, it'd be:
```
resp, err := someRequest()
if err == nil {
  resp.Close()
  t.Fatal("request foo succeeded, expected it to fail")
}
```

On the plus side, we protect ourselves from connection leaks.

fspmarshall reviewed (commented):
I think its best to reserve `nolint` for use in exceptional cases.  Seems like most of the cases here can be fixed with something like `if err == nil  { rsp.Body.Close() }`.

Also, instances of `nolint` usage should probably be accompanied by a note indicating *why* the lint should be ignored.

webvictim commented:
Agree with this.

Commit:
Enable agent forwarding with proxy recording mode

Agent forwarding is required to make proxy recording mode work.

Fixes #3606

awly commented:
@webvictim @fspmarshall @klizhentas @russjones PTAL

Fixed the integration test instead of skipping with TODO.
Removed all `nolint` comments except for one. That one linter violation is special and has a reason to remain. Left a comment explaining the situation.
